bson/unmarshal_test.go

```diff
@@ -166,3 +167,86 @@ func TestCachingDecodersNotSharedAcrossRegistries(t *testing.T) {
 		assert.Equal(t, int32(-1), *second.X, "expected X value to be -1, got %v", *second.X)
 	})
 }
+
+type expectedResponse struct {
+	DefinedField string
+}
+
+func unmarshalExtJSONHelper(extJSON string) (*expectedResponse, error) {
+	responseDoc := expectedResponse{}
+	err := UnmarshalExtJSON([]byte(extJSON), false, &responseDoc)
+	return &responseDoc, err
+}
+
+func TestUnmarshalExtJSONWithUndefinedField(t *testing.T) {
+	// When unmarshalling, fields that are undefined in the destination struct are skipped.
+	// This process must not skip other, defined fields and must not raise errors.
+	testCases := []struct {
+		name     string
+		testJSON string
+	}{
+		{
+			"no array",
+			`{
+				"UndefinedField": {"key": 1},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"outer array",
+			`{
+				"UndefinedField": [{"key": 1}],
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"embedded array",
+			`{
+				"UndefinedField": {"keys": [2]},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"outer array and embedded array",
+			`{
+				"UndefinedField": [{"keys": [2]}],
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"single embedded document",
+			`{
+				"UndefinedField": {"key": {"one": "two"}},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"doubly embedded document",
+			`{
+				"UndefinedField": {"key": {"one": {"two": "three"}}},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"embedded document and embedded array",
+			`{
+				"UndefinedField": {"key": {"one": {"two": [3]}}},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"embedded document and embedded array in outer array",
+			`{
+				"UndefinedField": [{"key": {"one": [3]}}],
+				"DefinedField": "value"
+			}`,
```

Comment:
Nice tests! Can we add cases for BSON values of type `CodeWithScope`, e.g. `{"$code": "foo", "$scope": {"bar": 1}}`? I only ask because `extJSONParser.peekType` and `extJSONParser.readValue` do handle it specially and it has an embedded document.

Also, in case you have not seen it, https://github.com/mongodb/specifications/blob/master/source/extended-json.rst#conversion-table has a nice reference for the forms of each BSON type in extended canonical JSON.

Reply:
Great catch! Added some `CodeWithScope` tests and a test for embedded documents with type definitions, as in `{"myDouble": {"$numberDouble": "1.24"}}`.